contracts/token/ERC20/ERC20Snapshot.sol

```diff
@@ -0,0 +1,130 @@
+    currentSnapshotId += 1;
+    emit Snapshot(currentSnapshotId);
+    return currentSnapshotId;
+  }
+
+  /**
+   * @dev Returns account balance for specific snapshot.
+   * @param account address The address to query the balance of.
+   * @param snapshotId uint256 The snapshot id for which to query the balance of.
+   * @return An uint256 representing the amount owned by the passed address for specific snapshot.
+   */
+  function balanceOfAt(
+    address account,
+    uint256 snapshotId
+  )
+    public
+    view
+    returns (uint256)
+  {
+    require(snapshotId > 0 && snapshotId <= currentSnapshotId);
+
+    uint256 idx = snapshotIds[account].findUpperBound(snapshotId);
+
+    if (idx == snapshotIds[account].length) {
+      return balanceOf(account);
+    } else {
+      return snapshotBalances[account][idx];
+    }
+  }
+
+  /**
+  * @dev Transfer token for a specified address. It takes balance snapshot for the sender and recipient account
+  * before transfer is done.
+  * @param to The address to transfer to.
+  * @param value The amount to be transferred.
+  */
+  function transfer(address to, uint256 value) public returns (bool) {
+    updateSnapshot(msg.sender);
+    updateSnapshot(to);
+    return super.transfer(to, value);
+  }
+
+  /**
+   * @dev Transfer tokens from one address to another. It takes balance snapshot of both accounts before
+   * the transfer is done.
+   * @param from address The address which you want to send tokens from
+   * @param to address The address which you want to transfer to
+   * @param value uint256 the amount of tokens to be transferred
+   */
+  function transferFrom(
+    address from,
+    address to,
+    uint256 value
+  ) 
+    public
+    returns (bool)
+  {
+    updateSnapshot(from);
+    updateSnapshot(to);
+    return super.transferFrom(from, to, value);
+  }
+
+  /**
+   * @dev Internal function that mints an amount of the token and assigns it to
+   * an account. This encapsulates the modification of balances such that the
+   * proper events are emitted. Takes snapshot before tokens are minted.
+   * @param account The account that will receive the created tokens.
+   * @param amount The amount that will be created.
+   */
+  function _mint(address account, uint256 amount) internal {
+    updateSnapshot(account);
+    super._mint(account, amount);
+  }
+
+  /**
+   * @dev Internal function that burns an amount of the token of a given
+   * account. Takes snapshot before tokens are burned.
+   * @param account The account whose tokens will be burnt.
+   * @param amount The amount that will be burnt.
+   */
+  function _burn(address account, uint256 amount) internal {
+    updateSnapshot(account);
+    super._burn(account, amount);
+  }
+
+  function updateSnapshot(address account) private {
+    if (lastSnapshotId(account) < currentSnapshotId) {
+      snapshotIds[account].push(currentSnapshotId);
+      snapshotBalances[account].push(balanceOf(account));
```

Comment:
`balanceOf(account)` will have the previous amount recorded. Perhaps this was a desired effect to have the history always behind and not current due to the nature of having to iterate a new snapshot manually. 

A discussion is to be had about having snapshots with an on/off feature or have snapshots that are always active.

Reply:
It's desired flow to have history always behind. If you want to have snapshots off then use regular ERC20.

Reply:
In your implementation, snapshots are off until another iteration is called upon. It is not an always on feature. 

I have to know which snapshot id I want to query, instead of using timestamp or blockNumber, you use a trivial number starting at 1 (0 == no snapshot exists yet). This will make it hard for user adoption as now I have to look up which snapshot id I want in relations to blockNumber/timestamp. 

Also, one needs to define which state of the balance do you care about at a specific block number? The balance before the block has been applied to the state or after?